Assertion error when freading data from a URL

```
>>> dt.fread("https://datasets.imdbws.com/title.basics.tsv.gz", fill=True)
100% |██████████████████████████████████████████████████| Downloading https://datasets.imdbws.com/title.basics.tsv.gz [done]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'end >= start' failed in src/core/read/parsers/parse_string.cc, line 47
```

- in release version of datatable this produces a seg.fault instead
- the dataset is a gzipped file, and fread currently cannot unzip files loaded from a URL. So it attempts to parse the underlying binary instead.

--- a/docs/releases/v1.0.0.rst
+++ b/docs/releases/v1.0.0.rst
@@ -15,6 +15,13 @@
       ``CASE WHEN`` in SQL. [#2656]
 
 
+    fread
+    -----
+
+    -[fix] Fix an error when reading a file with uneven number of fields
+      and having Windows-style newlines. [#2681]
+
+
     General
     -------
 

--- a/src/core/read/fread/fread_thread_context.cc
+++ b/src/core/read/fread/fread_thread_context.cc
@@ -98,7 +98,8 @@ void FreadThreadContext::read_chunk(
         tch++;
         j++;
       }
-      //*** END HOT. START TEPID ***//
+
+      const char* fieldEnd = tch;
       if (tch == tlineStart) {
         parse_ctx_.skip_whitespace_at_line_start();
         if (tch == parse_ctx_.eof) break;  // empty last line
@@ -109,13 +110,14 @@ void FreadThreadContext::read_chunk(
         parse_ctx_.target ++;
         j++;
         if (j==ncols) { used_nrows++; continue; }  // next line
-        tch--;
+        // TODO: fill the rest of the fields with NAs explicitly, without
+        //       going through type-bumping process below
+        tch = fieldEnd;
       }
       else {
         tch = fieldStart;
       }
     }
-    //*** END TEPID. NOW COLD.
 
     if (sep==' ') {
       while (tch < parse_ctx_.eof && *tch==' ') tch++;

--- a/src/core/read/parsers/parse_string.cc
+++ b/src/core/read/parsers/parse_string.cc
@@ -354,7 +354,7 @@ static void parse_string_naive(const ParseContext& ctx) {
     else if (static_cast<uint8_t>(c) <= 13) {  // probably a newline
       if (c == '\n') {
         // Move back to the beginning of \r+\n sequence
-        while (ch >= field_start && ch[-1] == '\r') ch--;
+        while (ch > field_start && ch[-1] == '\r') ch--;
         break;
       }
       if (c == '\r' && ctx.cr_is_newline) break;

--- a/tests/fread/test-fread-issues.py
+++ b/tests/fread/test-fread-issues.py
@@ -12,7 +12,7 @@ import pytest
 import random
 import re
 from datatable.internal import frame_integrity_check
-from tests import find_file, list_equals
+from tests import find_file, list_equals, assert_equals
 
 
 def test_issue1935():
@@ -541,3 +541,12 @@ def test_issue2523():
     # Make sure no AssertionError occurs here
     with pytest.raises(IOError):
         dt.fread("{\n  \"cells\": [\n    {\n\"import numpy \\n\",\n")
+
+
+def test_issue2681():
+    src = 'A,B,C\n' + 'rr,dd",g\n' + '1,2,3\n'*99 + 'abc,def\r\n'
+    DT = dt.fread(src, fill=True)
+    assert_equals(DT,
+        dt.Frame(A=['rr'] + ['1']*99 + ['abc'],
+                 B=['dd"'] + ['2']*99 + ['def'],
+                 C=['g'] + ['3']*99 + [None]))

